File: java/without_metaffi/call_python3_jep/src/test/java/BenchmarkTest.java
import jep.Interpreter;
import jep.SharedInterpreter;
import jep.JepException;
import org.junit.AfterClass;
import org.junit.BeforeClass;
import org.junit.Test;

import java.io.File;
import java.io.FileWriter;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import static org.junit.Assert.*;

/**
 * Performance benchmarks: Java host -> Python3 guest via Jep (embedded CPython).
 *
 * Uses Jep SharedInterpreter to call Python functions directly from Java.
 * Outputs results to tests/results/java_to_python3_jep.json.
 */
public class BenchmarkTest
{
	private static Interpreter interp;
	private static long interpStartupNs;

	private static int WARMUP;
	private static int ITERATIONS;

	@BeforeClass
	public static void setUp() throws Exception
	{
		String sourceRoot = System.getenv("METAFFI_SOURCE_ROOT");
		assertNotNull("METAFFI_SOURCE_ROOT must be set", sourceRoot);

		WARMUP = parseIntEnv("METAFFI_TEST_WARMUP", 100);
		ITERATIONS = parseIntEnv("METAFFI_TEST_ITERATIONS", 10000);

		// Start Jep interpreter and import guest module
		long startNs = System.nanoTime();

		interp = new SharedInterpreter();
		interp.exec("import sys");

		String modulePath = sourceRoot.replace('\\', '/') +
			"/sdk/test_modules/guest_modules/python3";
		interp.exec("sys.path.insert(0, '" + modulePath + "')");

		// Import guest module functions
		interp.exec("from module.core_functions import (wait_a_bit, div_integers, " +
			"join_strings, call_callback_add, returns_an_error)");
		interp.exec("from module.objects_and_classes import SomeClass, TestMap");
		interp.exec("from module.types_and_arrays import accepts_ragged_array");

		interpStartupNs = System.nanoTime() - startNs;
		System.err.println("Jep interpreter started (startup: " + interpStartupNs / 1_000_000 + " ms)");
	}

	@AfterClass
	public static void tearDown()
	{
		if (interp != null)
		{
			try { interp.close(); } catch (Exception ignored) {}
			interp = null;
		}
	}

	private static int parseIntEnv(String name, int defaultValue)
	{
		String val = System.getenv(name);
		if (val == null || val.isEmpty()) return defaultValue;
		return Integer.parseInt(val);
	}

	// ---- Statistical helpers ----

	private static double[] computeStats(long[] sortedNs)
	{
		int n = sortedNs.length;
		if (n == 0) return new double[]{0, 0, 0, 0, 0, 0, 0};

		double sum = 0;
		for (long v : sortedNs) sum += v;
		double mean = sum / n;

		double median = (n % 2 == 1) ? sortedNs[n / 2] : (sortedNs[n / 2 - 1] + sortedNs[n / 2]) / 2.0;
		double p95 = sortedNs[(int) (n * 0.95)];
		double p99 = sortedNs[Math.min((int) (n * 0.99), n - 1)];

		double sqDiffSum = 0;
		for (long v : sortedNs) sqDiffSum += (v - mean) * (v - mean);
		double stddev = Math.sqrt(sqDiffSum / n);
		double se = stddev / Math.sqrt(n);

		return new double[]{mean, median, p95, p99, stddev, mean - 1.96 * se, mean + 1.96 * se};
	}

	private static long[] removeOutliersIQR(long[] sortedNs)
	{
		int n = sortedNs.length;
		if (n < 4) return sortedNs;

		double q1 = sortedNs[n / 4];
		double q3 = sortedNs[3 * n / 4];
		double iqr = q3 - q1;
		double lower = q1 - 1.5 * iqr;
		double upper = q3 + 1.5 * iqr;

		List<Long> cleaned = new ArrayList<>();
		for (long v : sortedNs)
		{
			if (v >= lower && v <= upper) cleaned.add(v);
		}
		long[] result = new long[cleaned.size()];
		for (int i = 0; i < cleaned.size(); i++) result[i] = cleaned.get(i);
		return result;
	}

	private static long measureTimerOverhead()
	{
		long[] samples = new long[10000];
		for (int i = 0; i < 10000; i++)
		{
			long start = System.nanoTime();
			samples[i] = System.nanoTime() - start;
		}
		Arrays.sort(samples);
		return samples[5000];
	}

	// ---- Benchmark runner ----

	@FunctionalInterface
	interface BenchFn { void run() throws Throwable; }

	private static String runBenchmark(String scenario, Integer dataSize, int warmup, int iterations, BenchFn fn) throws Throwable
	{
		String label = scenario + (dataSize != null ? "[" + dataSize + "]" : "");
		System.err.println("  Benchmark: " + label + " (" + warmup + " warmup + " + iterations + " iterations)...");
		System.err.flush();

		for (int i = 0; i < warmup; i++)
		{
			try { fn.run(); }
			catch (Throwable e)
			{
				throw new RuntimeException("Benchmark '" + scenario + "' warmup iteration " + i + ": " + e.getMessage(), e);
			}
		}

		long[] rawNs = new long[iterations];
		for (int i = 0; i < iterations; i++)
		{
			long start = System.nanoTime();
			fn.run();
			rawNs[i] = System.nanoTime() - start;
		}

		long[] sortedNs = rawNs.clone();
		Arrays.sort(sortedNs);
		long[] cleaned = removeOutliersIQR(sortedNs);
		double[] stats = computeStats(cleaned);

		System.err.println("  Done: " + label + " (mean ~" + String.format("%.0f", stats[0]) + " ns)");

		StringBuilder sb = new StringBuilder();
		sb.append("    {\n");
		sb.append("      \"scenario\": \"").append(scenario).append("\",\n");
		sb.append("      \"data_size\": ").append(dataSize == null ? "null" : dataSize).append(",\n");
		sb.append("      \"status\": \"PASS\",\n");
		sb.append("      \"raw_iterations_ns\": [");
		for (int i = 0; i < rawNs.length; i++)
		{
			if (i > 0) sb.append(", ");
			sb.append(rawNs[i]);
		}
		sb.append("],\n");
		sb.append("      \"phases\": {\n");
		sb.append("        \"total\": {\n");
		sb.append("          \"mean_ns\": ").append(stats[0]).append(",\n");
		sb.append("          \"median_ns\": ").append(stats[1]).append(",\n");
		sb.append("          \"p95_ns\": ").append(stats[2]).append(",\n");
		sb.append("          \"p99_ns\": ").append(stats[3]).append(",\n");
		sb.append("          \"stddev_ns\": ").append(stats[4]).append(",\n");
		sb.append("          \"ci95_ns\": [").append(stats[5]).append(", ").append(stats[6]).append("]\n");
		sb.append("        }\n");
		sb.append("      }\n");
		sb.append("    }");
		return sb.toString();
	}

	private static String makeFailedResult(String scenario, Integer dataSize, String error)
	{
		StringBuilder sb = new StringBuilder();
		sb.append("    {\n");
		sb.append("      \"scenario\": \"").append(scenario).append("\",\n");
		sb.append("      \"data_size\": ").append(dataSize == null ? "null" : dataSize).append(",\n");
		sb.append("      \"status\": \"FAIL\",\n");
		sb.append("      \"error\": \"").append(error.replace("\"", "\\\"").replace("\n", "\\n")).append("\",\n");
		sb.append("      \"raw_iterations_ns\": [],\n");
		sb.append("      \"phases\": {}\n");
		sb.append("    }");
		return sb.toString();
	}

	// ---- Main benchmark test ----

	@Test
	public void testAllBenchmarks() throws Throwable
	{
		String mode = System.getenv().getOrDefault("METAFFI_TEST_MODE", "");
		if ("correctness".equals(mode))
		{
			System.err.println("Skipping benchmarks: METAFFI_TEST_MODE=correctness");
			return;
		}

		long timerOverhead = measureTimerOverhead();
		System.err.println("Timer overhead: " + timerOverhead + " ns");

		List<String> benchmarkJsons = new ArrayList<>();

		// --- Scenario 1: Void call ---
		benchmarkJsons.add(runBenchmark("void_call", null, WARMUP, ITERATIONS,
			() -> interp.exec("wait_a_bit(0)")));

		// --- Scenario 2: Primitive echo ---
		benchmarkJsons.add(runBenchmark("primitive_echo", null, WARMUP, ITERATIONS,
			() -> {
				interp.exec("_r = div_integers(10, 2)");
				Object result = interp.getValue("_r");
				double val = ((Number) result).doubleValue();
				if (Math.abs(val - 5.0) > 1e-10)
				{
					throw new RuntimeException("div_integers: got " + val + ", want 5.0");
				}
			}));

		// --- Scenario 3: String echo ---
		benchmarkJsons.add(runBenchmark("string_echo", null, WARMUP, ITERATIONS,
			() -> {
				interp.exec("_r = join_strings(['hello', 'world'])");
				String result = (String) interp.getValue("_r");
				if (!"hello,world".equals(result))
				{
					throw new RuntimeException("join_strings: got " + result);
				}
			}));

		// --- Scenario 4: Array sum (varying sizes) ---
		for (int size : new int[]{10, 100, 1000, 10000})
		{
			// Pre-create the Python list for this size
			interp.exec("_arr_" + size + " = [[i+1 for i in range(" + size + ")]]");
			long expectedSum = (long) size * (size + 1) / 2;
			final long expected = expectedSum;
			final String arrName = "_arr_" + size;

			benchmarkJsons.add(runBenchmark("array_sum", size, WARMUP, ITERATIONS,
				() -> {
					interp.exec("_r = accepts_ragged_array(" + arrName + ")");
					Object result = interp.getValue("_r");
					long val = ((Number) result).longValue();
					if (val != expected)
					{
						throw new RuntimeException("array_sum: got " + val + ", want " + expected);
					}
				}));
		}

		// --- Scenario 5: Object method ---
		benchmarkJsons.add(runBenchmark("object_method", null, WARMUP, ITERATIONS,
			() -> {
				interp.exec("_obj = SomeClass('bench')");
				interp.exec("_r = _obj.print()");
				String result = (String) interp.getValue("_r");
				if (result == null || !result.contains("bench"))
				{
					throw new RuntimeException("SomeClass.print: got " + result);
				}
			}));

		// --- Scenario 6: Callback ---
		try
		{
			// Define a Python callback function that adds two numbers
			interp.exec("def _java_add(a, b): return a + b");

			benchmarkJsons.add(runBenchmark("callback", null, WARMUP, ITERATIONS,
				() -> {
					interp.exec("_r = call_callback_add(_java_add)");
					Object result = interp.getValue("_r");
					long val = ((Number) result).longValue();
					if (val != 3L)
					{
						throw new RuntimeException("call_callback_add: got " + val + ", want 3");
					}
				}));
		}
		catch (Throwable e)
		{
			System.err.println("Callback scenario failed: " + e.getMessage());
			benchmarkJsons.add(makeFailedResult("callback", null, e.getMessage()));
		}

		// --- Scenario 7: Error propagation ---
		benchmarkJsons.add(runBenchmark("error_propagation", null, WARMUP, ITERATIONS,
			() -> {
				try
				{
					interp.exec("returns_an_error()");
					throw new RuntimeException("returns_an_error did not throw");
				}
				catch (JepException e)
				{
					// Expected: Python error -> JepException
				}
			}));

		// --- Write results ---
		writeResults(benchmarkJsons, timerOverhead);
	}

	private void writeResults(List<String> benchmarkJsons, long timerOverhead)
	{
		String resultPath = System.getenv().getOrDefault("METAFFI_TEST_RESULTS_FILE", "");
		if (resultPath == null || resultPath.isEmpty())
		{
			String sourceRoot = System.getenv("METAFFI_SOURCE_ROOT");
			resultPath = sourceRoot + "/tests/results/java_to_python3_jep.json";
		}

		StringBuilder sb = new StringBuilder();
		sb.append("{\n");
		sb.append("  \"metadata\": {\n");
		sb.append("    \"host\": \"java\",\n");
		sb.append("    \"guest\": \"python3\",\n");
		sb.append("    \"mechanism\": \"jep\",\n");
		sb.append("    \"timestamp\": \"").append(java.time.Instant.now().toString()).append("\",\n");
		sb.append("    \"environment\": {\n");
		sb.append("      \"os\": \"").append(System.getProperty("os.name").toLowerCase()).append("\",\n");
		sb.append("      \"arch\": \"").append(System.getProperty("os.arch")).append("\",\n");
		sb.append("      \"java_version\": \"").append(System.getProperty("java.version")).append("\"\n");
		sb.append("    },\n");
		sb.append("    \"config\": {\n");
		sb.append("      \"warmup_iterations\": ").append(WARMUP).append(",\n");
		sb.append("      \"measured_iterations\": ").append(ITERATIONS).append(",\n");
		sb.append("      \"timer_overhead_ns\": ").append(timerOverhead).append("\n");
		sb.append("    }\n");
		sb.append("  },\n");
		sb.append("  \"initialization\": {\n");
		sb.append("    \"interpreter_startup_ns\": ").append(interpStartupNs).append("\n");
		sb.append("  },\n");
		sb.append("  \"correctness\": null,\n");
		sb.append("  \"benchmarks\": [\n");

		for (int i = 0; i < benchmarkJsons.size(); i++)
		{
			if (i > 0) sb.append(",\n");
			sb.append(benchmarkJsons.get(i));
		}

		sb.append("\n  ]\n");
		sb.append("}\n");

		try
		{
			File file = new File(resultPath);
			file.getParentFile().mkdirs();
			try (PrintWriter pw = new PrintWriter(new FileWriter(file)))
			{
				pw.print(sb.toString());
			}
			System.err.println("Results written to " + resultPath);
		}
		catch (Exception e)
		{
			System.err.println("Failed to write results: " + e.getMessage());
			fail("Failed to write benchmark results: " + e.getMessage());
		}
	}
}

File: go/without_metaffi/call_python3_grpc/benchmark_test.go
package call_python3_grpc

import (
	"bufio"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"math"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"sort"
	"strings"
	"testing"
	"time"

	pb "github.com/MetaFFI/tests/go/without_metaffi/call_python3_grpc/pb"
	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials/insecure"
)

// ---------------------------------------------------------------------------
// Global state
// ---------------------------------------------------------------------------

var (
	client     pb.BenchmarkServiceClient
	conn       *grpc.ClientConn
	serverCmd  *exec.Cmd
	serverAddr string

	// Init timing
	serverStartNs int64
	connectNs     int64
)

// ---------------------------------------------------------------------------
// TestMain -- start Python gRPC server, connect, run tests, stop server
// ---------------------------------------------------------------------------

func TestMain(m *testing.M) {
	srcRoot := os.Getenv("METAFFI_SOURCE_ROOT")
	if srcRoot == "" {
		fmt.Fprintln(os.Stderr, "FATAL: METAFFI_SOURCE_ROOT must be set")
		os.Exit(1)
	}

	modulePath := filepath.Join(srcRoot, "sdk", "test_modules", "guest_modules", "python3")

	// Determine Python executable
	pythonExe := os.Getenv("PYTHON_EXE")
	if pythonExe == "" {
		pythonExe = "python"
	}

	// Resolve server.py path relative to the test directory
	serverScript := filepath.Join("server", "server.py")

	// --- Start Python gRPC server ---
	startTime := time.Now()

	serverCmd = exec.Command(pythonExe, serverScript, "--module-path", modulePath)
	serverCmd.Dir = getTestDir()
	serverCmd.Stderr = os.Stderr

	stdout, err := serverCmd.StdoutPipe()
	if err != nil {
		fmt.Fprintf(os.Stderr, "FATAL: failed to get stdout pipe: %v\n", err)
		os.Exit(1)
	}

	if err := serverCmd.Start(); err != nil {
		fmt.Fprintf(os.Stderr, "FATAL: failed to start gRPC server: %v\n", err)
		os.Exit(1)
	}

	// Read "READY:<port>" from stdout
	scanner := bufio.NewScanner(stdout)
	if !scanner.Scan() {
		serverCmd.Process.Kill()
		fmt.Fprintln(os.Stderr, "FATAL: gRPC server did not print READY line")
		os.Exit(1)
	}

	line := scanner.Text()
	if !strings.HasPrefix(line, "READY:") {
		serverCmd.Process.Kill()
		fmt.Fprintf(os.Stderr, "FATAL: unexpected server output: %q\n", line)
		os.Exit(1)
	}
	port := strings.TrimPrefix(line, "READY:")
	serverAddr = "localhost:" + port
	serverStartNs = time.Since(startTime).Nanoseconds()

	// Drain remaining stdout in background
	go func() { io.Copy(io.Discard, stdout) }()

	// --- Connect to server ---
	connectStart := time.Now()
	conn, err = grpc.NewClient(
		serverAddr,
		grpc.WithTransportCredentials(insecure.NewCredentials()),
	)
	if err != nil {
		serverCmd.Process.Kill()
		fmt.Fprintf(os.Stderr, "FATAL: failed to connect to gRPC server at %s: %v\n", serverAddr, err)
		os.Exit(1)
	}
	client = pb.NewBenchmarkServiceClient(conn)

	// Verify connectivity with a ping
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	_, pingErr := client.VoidCall(ctx, &pb.VoidCallRequest{Secs: 0})
	cancel()
	if pingErr != nil {
		conn.Close()
		serverCmd.Process.Kill()
		fmt.Fprintf(os.Stderr, "FATAL: gRPC server not responding: %v\n", pingErr)
		os.Exit(1)
	}
	connectNs = time.Since(connectStart).Nanoseconds()

	// --- Run tests ---
	code := m.Run()

	// --- Cleanup ---
	conn.Close()
	serverCmd.Process.Kill()
	serverCmd.Wait()
	os.Exit(code)
}

// getTestDir returns the directory containing the test files.
func getTestDir() string {
	// When running `go test`, CWD is the package directory
	dir, err := os.Getwd()
	if err != nil {
		return "."
	}
	return dir
}

// ---------------------------------------------------------------------------
// Benchmark configuration
// ---------------------------------------------------------------------------

func getIntEnv(key string, defaultVal int) int {
	v := os.Getenv(key)
	if v == "" {
		return defaultVal
	}
	var n int
	_, err := fmt.Sscanf(v, "%d", &n)
	if err != nil {
		return defaultVal
	}
	return n
}

// ---------------------------------------------------------------------------
// Statistics helpers (same schema as other benchmarks)
// ---------------------------------------------------------------------------

type PhaseStats struct {
	MeanNs   float64    `json:"mean_ns"`
	MedianNs float64    `json:"median_ns"`
	P95Ns    float64    `json:"p95_ns"`
	P99Ns    float64    `json:"p99_ns"`
	StddevNs float64    `json:"stddev_ns"`
	CI95Ns   [2]float64 `json:"ci95_ns"`
}

type BenchmarkResult struct {
	Scenario        string                `json:"scenario"`
	DataSize        *int                  `json:"data_size"`
	Status          string                `json:"status"`
	RawIterationsNs []int64               `json:"raw_iterations_ns"`
	Phases          map[string]PhaseStats `json:"phases"`
}

type ResultFile struct {
	Metadata    Metadata          `json:"metadata"`
	Correctness interface{}       `json:"correctness"`
	Init        InitTiming        `json:"initialization"`
	Benchmarks  []BenchmarkResult `json:"benchmarks"`
}

type Metadata struct {
	Host        string      `json:"host"`
	Guest       string      `json:"guest"`
	Mechanism   string      `json:"mechanism"`
	Timestamp   string      `json:"timestamp"`
	Environment Environment `json:"environment"`
	Config      Config      `json:"config"`
}

type Environment struct {
	OS        string `json:"os"`
	Arch      string `json:"arch"`
	GoVersion string `json:"go_version"`
}

type Config struct {
	WarmupIterations   int   `json:"warmup_iterations"`
	MeasuredIterations int   `json:"measured_iterations"`
	TimerOverheadNs    int64 `json:"timer_overhead_ns"`
}

type InitTiming struct {
	ServerStartNs int64 `json:"server_start_ns"`
	ConnectNs     int64 `json:"connect_ns"`
}

func computeStats(sorted []int64) PhaseStats {
	n := len(sorted)
	if n == 0 {
		return PhaseStats{}
	}

	var sum float64
	for _, v := range sorted {
		sum += float64(v)
	}
	mean := sum / float64(n)

	var median float64
	if n%2 == 0 {
		median = float64(sorted[n/2-1]+sorted[n/2]) / 2.0
	} else {
		median = float64(sorted[n/2])
	}

	p95 := float64(sorted[int(float64(n)*0.95)])
	p99 := float64(sorted[int(math.Min(float64(n)*0.99, float64(n-1)))])

	var sqDiffSum float64
	for _, v := range sorted {
		diff := float64(v) - mean
		sqDiffSum += diff * diff
	}
	stddev := math.Sqrt(sqDiffSum / float64(n))

	se := stddev / math.Sqrt(float64(n))

	return PhaseStats{
		MeanNs:   mean,
		MedianNs: median,
		P95Ns:    p95,
		P99Ns:    p99,
		StddevNs: stddev,
		CI95Ns:   [2]float64{mean - 1.96*se, mean + 1.96*se},
	}
}

func removeOutliersIQR(sorted []int64) []int64 {
	n := len(sorted)
	if n < 4 {
		return sorted
	}

	q1 := float64(sorted[n/4])
	q3 := float64(sorted[3*n/4])
	iqr := q3 - q1
	lower := q1 - 1.5*iqr
	upper := q3 + 1.5*iqr

	result := make([]int64, 0, n)
	for _, v := range sorted {
		if float64(v) >= lower && float64(v) <= upper {
			result = append(result, v)
		}
	}
	return result
}

func measureTimerOverhead() int64 {
	const n = 10000
	samples := make([]int64, n)
	for i := 0; i < n; i++ {
		start := time.Now()
		elapsed := time.Since(start)
		samples[i] = elapsed.Nanoseconds()
	}
	sort.Slice(samples, func(i, j int) bool { return samples[i] < samples[j] })
	return samples[n/2]
}

// ---------------------------------------------------------------------------
// Benchmark runner
// ---------------------------------------------------------------------------

func runBenchmark(
	t *testing.T,
	scenario string,
	dataSize *int,
	warmup int,
	iterations int,
	benchFn func() error,
) BenchmarkResult {
	t.Helper()

	for i := 0; i < warmup; i++ {
		if err := benchFn(); err != nil {
			t.Fatalf("benchmark %q warmup iteration %d: %v", scenario, i, err)
			return BenchmarkResult{Scenario: scenario, DataSize: dataSize, Status: "FAIL"}
		}
	}

	rawNs := make([]int64, iterations)
	for i := 0; i < iterations; i++ {
		start := time.Now()
		err := benchFn()
		elapsed := time.Since(start).Nanoseconds()

		if err != nil {
			t.Fatalf("benchmark %q iteration %d: %v (BENCHMARK INVALIDATED)", scenario, i, err)
			return BenchmarkResult{Scenario: scenario, DataSize: dataSize, Status: "FAIL"}
		}
		rawNs[i] = elapsed
	}

	sortedNs := make([]int64, len(rawNs))
	copy(sortedNs, rawNs)
	sort.Slice(sortedNs, func(i, j int) bool { return sortedNs[i] < sortedNs[j] })

	cleaned := removeOutliersIQR(sortedNs)
	totalStats := computeStats(cleaned)

	return BenchmarkResult{
		Scenario:        scenario,
		DataSize:        dataSize,
		Status:          "PASS",
		RawIterationsNs: rawNs,
		Phases: map[string]PhaseStats{
			"total": totalStats,
		},
	}
}

// ---------------------------------------------------------------------------
// Benchmark tests (7 scenarios)
// ---------------------------------------------------------------------------

func TestBenchmarkAll(t *testing.T) {
	mode := os.Getenv("METAFFI_TEST_MODE")
	if mode == "correctness" {
		t.Skip("Skipping benchmarks: METAFFI_TEST_MODE=correctness")
	}

	warmup := getIntEnv("METAFFI_TEST_WARMUP", 100)
	iterations := getIntEnv("METAFFI_TEST_ITERATIONS", 10000)

	timerOverhead := measureTimerOverhead()
	t.Logf("Timer overhead: %d ns", timerOverhead)

	var benchmarks []BenchmarkResult

	// --- Scenario 1: Void call ---
	t.Run("void_call", func(t *testing.T) {
		result := runBenchmark(t, "void_call", nil, warmup, iterations, func() error {
			_, err := client.VoidCall(context.Background(), &pb.VoidCallRequest{Secs: 0})
			return err
		})
		benchmarks = append(benchmarks, result)
	})

	// --- Scenario 2: Primitive echo ---
	t.Run("primitive_echo", func(t *testing.T) {
		result := runBenchmark(t, "primitive_echo", nil, warmup, iterations, func() error {
			resp, err := client.DivIntegers(context.Background(), &pb.DivIntegersRequest{X: 10, Y: 2})
			if err != nil {
				return err
			}
			if math.Abs(resp.Result-5.0) > 1e-10 {
				return fmt.Errorf("div_integers: got %v, want 5.0", resp.Result)
			}
			return nil
		})
		benchmarks = append(benchmarks, result)
	})

	// --- Scenario 3: String echo ---
	t.Run("string_echo", func(t *testing.T) {
		result := runBenchmark(t, "string_echo", nil, warmup, iterations, func() error {
			resp, err := client.JoinStrings(context.Background(), &pb.JoinStringsRequest{
				Values: []string{"hello", "world"},
			})
			if err != nil {
				return err
			}
			if resp.Result != "hello,world" {
				return fmt.Errorf("join_strings: got %q, want \"hello,world\"", resp.Result)
			}
			return nil
		})
		benchmarks = append(benchmarks, result)
	})

	// --- Scenario 4: Array sum (varying sizes) ---
	for _, size := range []int{10, 100, 1000, 10000} {
		size := size
		t.Run(fmt.Sprintf("array_sum_%d", size), func(t *testing.T) {
			// Pre-build the values slice
			values := make([]int64, size)
			var expectedSum int64
			for i := 0; i < size; i++ {
				values[i] = int64(i + 1)
				expectedSum += int64(i + 1)
			}

			sizePtr := size
			result := runBenchmark(t, "array_sum", &sizePtr, warmup, iterations, func() error {
				resp, err := client.ArraySum(context.Background(), &pb.ArraySumRequest{
					Values: values,
				})
				if err != nil {
					return err
				}
				if resp.Sum != expectedSum {
					return fmt.Errorf("array_sum: got %d, want %d", resp.Sum, expectedSum)
				}
				return nil
			})
			benchmarks = append(benchmarks, result)
		})
	}

	// --- Scenario 5: Object create + method call ---
	t.Run("object_method", func(t *testing.T) {
		result := runBenchmark(t, "object_method", nil, warmup, iterations, func() error {
			resp, err := client.ObjectMethod(context.Background(), &pb.ObjectMethodRequest{
				Name: "bench",
			})
			if err != nil {
				return err
			}
			if resp.Result != "Hello from SomeClass bench" {
				return fmt.Errorf("object_method: got %q, want \"Hello from SomeClass bench\"", resp.Result)
			}
			return nil
		})
		benchmarks = append(benchmarks, result)
	})

	// --- Scenario 6: Callback (bidirectional streaming) ---
	t.Run("callback", func(t *testing.T) {
		result := runBenchmark(t, "callback", nil, warmup, iterations, func() error {
			stream, err := client.CallbackAdd(context.Background())
			if err != nil {
				return fmt.Errorf("open stream: %w", err)
			}

			// Send invoke
			if err := stream.Send(&pb.CallbackClientMsg{
				Msg: &pb.CallbackClientMsg_Invoke{Invoke: true},
			}); err != nil {
				return fmt.Errorf("send invoke: %w", err)
			}

			// Receive compute request
			resp, err := stream.Recv()
			if err != nil {
				return fmt.Errorf("recv compute: %w", err)
			}
			compute := resp.GetCompute()
			if compute == nil {
				return fmt.Errorf("expected compute message, got %v", resp)
			}

			// Compute and send result
			sum := compute.A + compute.B
			if err := stream.Send(&pb.CallbackClientMsg{
				Msg: &pb.CallbackClientMsg_AddResult{AddResult: sum},
			}); err != nil {
				return fmt.Errorf("send result: %w", err)
			}

			// Receive final result
			resp, err = stream.Recv()
			if err != nil {
				return fmt.Errorf("recv final: %w", err)
			}
			if resp.GetFinalResult() != 3 {
				return fmt.Errorf("callback: got %d, want 3", resp.GetFinalResult())
			}

			stream.CloseSend()
			return nil
		})
		benchmarks = append(benchmarks, result)
	})

	// --- Scenario 7: Error propagation ---
	t.Run("error_propagation", func(t *testing.T) {
		result := runBenchmark(t, "error_propagation", nil, warmup, iterations, func() error {
			_, err := client.ReturnsAnError(context.Background(), &pb.Empty{})
			if err == nil {
				return fmt.Errorf("expected gRPC error but got nil")
			}
			// Error IS expected -- this is the successful path
			return nil
		})
		benchmarks = append(benchmarks, result)
	})

	// --- Write results ---
	writeResults(t, benchmarks, timerOverhead, warmup, iterations)
}

// ---------------------------------------------------------------------------
// JSON output
// ---------------------------------------------------------------------------

func writeResults(t *testing.T, benchmarks []BenchmarkResult, timerOverhead int64, warmup, iterations int) {
	t.Helper()

	resultPath := os.Getenv("METAFFI_TEST_RESULTS_FILE")
	if resultPath == "" {
		resultPath = filepath.Join("..", "..", "..", "results", "go_to_python3_grpc.json")
	}

	result := ResultFile{
		Metadata: Metadata{
			Host:      "go",
			Guest:     "python3",
			Mechanism: "grpc",
			Timestamp: time.Now().UTC().Format(time.RFC3339),
			Environment: Environment{
				OS:        runtime.GOOS,
				Arch:      runtime.GOARCH,
				GoVersion: runtime.Version(),
			},
			Config: Config{
				WarmupIterations:   warmup,
				MeasuredIterations: iterations,
				TimerOverheadNs:    timerOverhead,
			},
		},
		Init: InitTiming{
			ServerStartNs: serverStartNs,
			ConnectNs:     connectNs,
		},
		Benchmarks: benchmarks,
	}

	data, err := json.MarshalIndent(result, "", "  ")
	if err != nil {
		t.Fatalf("Failed to marshal results to JSON: %v", err)
	}

	if err := os.WriteFile(resultPath, data, 0644); err != nil {
		t.Logf("WARNING: Failed to write results to %s: %v", resultPath, err)
	} else {
		t.Logf("Results written to %s", resultPath)
	}
}
